### PR TITLE
[JIT] Split backend API between libtorch and libtorch_python

### PR DESCRIPTION
**Pull request by SplitInfinity**
**Summary**
This commit splits the code for the JIT backend API code between
`libtorch` and `libtorch_python` so that the code required to lower a
Module for a backend is in `libtorch_python` and the code required to
run a lowered Module is in `libtorch`. This allows a lowered `Module` to
be imported and executed inside a Python-less environment (provided that
the definition for the backend class is included).

Test Plan:
`python test/test_jit.py TestBackends`

### CODE REVIEW

**Comment by SplitInfinity** [comment]
I think that this (splitting the API between `libtorch_python` and `libtorch`) needs to be done, but I am not 100% happy with this approach because the user has to get the name of the backend right twice:
```
  static auto constexpr backend_name = "test_backend";
  static auto cls = torch::jit::backend<TestBackend>(backend_name);
  auto m = py::handle(module).cast<py::module>();
  m.def("_jit_to_test_backend", testBackend().generateToBackendFn());
  m.def("_jit_to_test_backend", torch::jit::generateToBackendFn(backend_name));
```

Ideally, I would have liked to declare `generateToBackendFn` in `backend.h` and define it in `backend.cpp`, but this is not possible because `torch::jit::backend` is templated. Even though `generateToBackendFn` is not templated, it must be defined in the header (hopefully I am wrong about this).

File: torch/csrc/jit/backends/backend.h [hunk 1]
```diff
@@ -1,13 +1,7 @@
...
 
-#include <ATen/core/builtin_function.h>
-#include <ATen/core/stack.h>
 #include <torch/csrc/jit/backends/backend_detail.h>
 #include <torch/csrc/jit/backends/backend_interface.h>
-#include <torch/csrc/jit/backends/backend_resolver.h>
-#include <torch/csrc/jit/frontend/code_template.h>
-#include <torch/csrc/jit/frontend/resolver.h>
-#include <torch/csrc/jit/frontend/sugared_value.h>
 #include <torch/csrc/jit/python/pybind_utils.h>
```
Comment (suo) [inline torch/csrc/jit/backends/backend.h:5]: I don't think we can use this here, otherwise this header is not includable without libtorch_python, right?

**Review by suo — CHANGES_REQUESTED** [review]
I think we need to separate things more cleanly here.
- `backend.h`, `backend_detail.h` and `backend.cpp`: all the cpp-only stuff
- `python_backend.h` and `python_backend.cpp` (or a different name): what you include if you're writing the pybinding.

Right now it seems a little mixed up (including `backend.h` includes `pybind_utils.h`)

Also, I think we need a test like the ones for custom ops and custom classes that ensure that this e2e flow works. Those basically define a custom op in a library, then build a small binary that links in only libtorch and that library and tries to use the op.

**Comment by SplitInfinity** [comment]
See #40253.

**Pull request closed without merging**